content/200-concepts/100-components/02-prisma-client/061-custom-validation.mdx

```diff
@@ -6,12 +6,119 @@ metaDescription: 'This page explains how to add custom validation to Prisma Clie
+- [Prisma Client extensions](/concepts/components/prisma-client/client-extensions)
+- A custom function
+
+You can use any validation library you'd like. The Node.js ecosystem offers a number of high-quality, easy-to-use validation libraries to choose from including: [joi](https://github.com/sideway/joi), [validator.js](https://github.com/validatorjs/validator.js), [Yup](https://github.com/jquense/yup), [Zod](https://github.com/colinhacks/zod) and [Superstruct](https://github.com/ianstormtaylor/superstruct).
 
 </TopBlock>
 
+## Input validation with Prisma Client extensions
+
+> Prisma Client extensions are currently in [Preview](/about/prisma/releases#preview).
+
+This example adds runtime validation when creating and updating values using a Zod schema to check that the data passed to Prisma Client is valid.
+
+<Admonition>
+
+Query extensions do not currently work for nested operations. In this example, validations are only run on the top level data object passed to methods such as `prisma.product.create()`. Validations implemented this way do not automatically run for [nested writes](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes).
+
+</Admonition>
+
+<TabbedContent tabs={[ <FileWithIcon text="Prisma Client extension" icon="file"/>,<FileWithIcon text="Example query" icon="file"/>, <FileWithIcon text="Prisma schema" icon="file"/>]}>
+
+<tab>
+
+```ts copy
+import { PrismaClient } from "@prisma/client";
+import { z } from 'zod'
+import { type Prisma } from '@prisma/client'
+
+const PostCreateInput = z.object({
+  title: z.string(),
+  published: z.boolean().default(false),
+  content: z.string().nullable(),
+  authorId: z.number().nullable(),
+}) satisfies z.Schema<Prisma.PostUncheckedCreateInput>
+
+const prisma = new PrismaClient().$extends({
+  query: {
+    post: {
+      create({ args, query }) {
+        args.data = PostCreateInput.parse(args.data)
+        return query(args)
+      },
+      update({ args, query }) {
+        args.data = PostCreateInput.partial().parse(args.data)
+        return query(args)
+      },
+      updateMany({ args, query }) {
+        args.data = PostCreateInput.partial().parse(args.data)
+        return query(args)
+      },
+      upsert({ args, query }) {
+        args.create = PostCreateInput.parse(args.create)
+        args.update = PostCreateInput.parse(args.update)
+        return query(args)
+      }
+    }
+  }
+})
+```
+
+</tab>
+
+<tab>
+
+```ts copy
+// A post with invalid input — missing the title field
+const post = await prisma.post.create({
+  data: {
+    content: 'Post content',
+  },
+})
+```
+
+</tab>
+
+<tab>
+
+```prisma file="prisma/schema.prisma" copy
```

Comment:
the extra file name looks a bit glitchy when used in combination with tabs: 

<img width="513" alt="image" src="https://user-images.githubusercontent.com/4058327/234212765-f708f5fe-432d-411e-b3b7-aa48a6512544.png">


```suggestion
```prisma copy
```